src/libaktualizr/package_manager/packagemanagerinterface.h

```diff
@@ -35,12 +35,16 @@ enum class TargetStatus {
 
 class PackageManagerInterface {
  public:
-  PackageManagerInterface(PackageConfig pconfig, BootloaderConfig bconfig, std::shared_ptr<INvStorage> storage,
-                          std::shared_ptr<HttpInterface> http)
+  PackageManagerInterface(const PackageConfig& pconfig, const BootloaderConfig& bconfig,
+                          const std::shared_ptr<INvStorage>& storage, const std::shared_ptr<HttpInterface>& http)
+      : PackageManagerInterface(pconfig, std::make_shared<Bootloader>(bconfig, storage), storage, http) {}
+
+  PackageManagerInterface(PackageConfig pconfig, std::shared_ptr<Bootloader> bootloader,
+                          std::shared_ptr<INvStorage> storage, std::shared_ptr<HttpInterface> http)
       : config(std::move(pconfig)),
         storage_(std::move(storage)),
         http_(std::move(http)),
-        bootloader_{new Bootloader(std::move(bconfig), *storage_)} {}
+        bootloader_(std::move(bootloader)) {}
```

Comment:
I didn't change it back, I just added an additional delegating ctor since I needed to pass a bootloader instance created by a client/app of a package manager. BTW, I can imagine a package manager interface implementation that does not require any bootloader/bootloader-config (e.g. file/docker conf update), so putting it into the package manager interface is questionable, actually all params of the packman interface ctor should be removed.

Reply:
> I didn't change it back, I just added an additional delegating ctor since I needed to pass a bootloader instance created by a client/app of a package manager.

Ah yes, now I see what you mean. Sorry, carry on!

> I can imagine a package manager interface implementation that does not require any bootloader/bootloader-config (e.g. file/docker conf update), so putting it into the package manager interface is questionable, actually all params of the packman interface ctor should be removed.

I can imagine that, too, so perhaps we should have a conversation with @eu-smirnov and others once we're all available to clear up our plan here.

Reply:
What makes it necessary to create Bootloader outside of PackageManager instance? What I wanted to do next here is to merge bootloader and package manager configurations and to remove this parameter completely. Currently Bootloader functionality is used by Ostree manager only.

Reply:
> What makes it necessary to create Bootloader outside of PackageManager instance?
it's used in the secondary implementation (AktualizrSecondary class).

> and to remove this parameter completely
I agree, moreover I think, it has to be removed from PackageManagerInterface class at all as well as all other parameters of its ctor to make it a proper interface.

> Currently Bootloader functionality is used by Ostree manager only.

It's not so, it's used in PackageManagerInterface and PackageManagerFake implementations too in order to implement a few PackageManagerInterface methods. I actually think that Bootloader should be moved out of PackageManagerInterface, PackageManagerFake and OstreeManager at all as well as those methods from PackageManagerInterface. I think, the bootloader should be owned by Primary (sotauptaneclient) and Secondary (AktualizrSecondary) as the bootloading functionality is not package/update specific it's rather ECU/system specific functionality.

Reply:
> It's not so, it's used in PackageManagerInterface and PackageManagerFake implementations too in order to implement a few PackageManagerInterface methods. I actually think that Bootloader should be moved out of PackageManagerInterface, PackageManagerFake and OstreeManager at all as well as those methods from PackageManagerInterface. I think, the bootloader should be owned by Primary (sotauptaneclient) and Secondary (AktualizrSecondary) as the bootloading functionality is not package/update specific it's rather ECU/system specific functionality.

Those calls are basically bogus, the only place where it's *really* used is together with ostree to provide rollback. So to me it looks like functionality specific to package manager. Except, maybe `reboot`, which shouldn't be a part of libaktuaizr at all.

Reply:
> it's really used is together with ostree to provide rollback

I don't see any place in the ostree package manager implementation where the instance of bootloader is used for the rollback. The three calls to the bootloader that the ostree pacman makes are not directly related to actual rollback and they can be made from the sotauptane client.

The bootloader implementation is not package management specific (does it have any ostree specifics ? ) and can be reused for more than one package manager type, I can imagine another (non-ostree) type that require some actions after reboot too.

Having said that, I see two options here:

The first one when we can have a single instance of a bootloader in a scope of the sotauptane client and do `if (!bootloader_->rebootDetected()) {` in finalizeAfterReboot() instead of `if (!package_manager_->rebootDetected()) {` and we can easily remove the bootloader instance from the ostree pacman at all;

The second one is to reuse the bootloader class in each package manager that requires it like in case of the ostree pacman. The problem is that each package manager class should implement the reboot related methods in this case in order to make sotauptaneclient generic (`if (!package_manager_->rebootDetected())`).

Reply:
> The first one when we can have a single instance of a bootloader in a scope of the sotauptane client and do `if (!bootloader_->rebootDetected()) {` in finalizeAfterReboot() instead of `if (!package_manager_->rebootDetected()) {` and we can easily remove the bootloader instance from the ostree pacman at all;
> 
> The second one is to reuse the bootloader class in each package manager that requires it like in case of the ostree pacman. The problem is that each package manager class should implement the reboot related methods in this case in order to make sotauptaneclient generic (`if (!package_manager_->rebootDetected())`).

I don't think this logic belongs to SotaUptaneClient. What SotaUptaneClient really cares about is that it has a pending target, and it can ask PackageManager to try to finalize the installation. Three outcomes are possible: finalization success, failure, and preconditions for finalization are not met (e.g. `!rebootDetected()`). Things like `rebootDetected()` and `rebootFlagClear()` should be an internal business of the PackageManager, as it's the only entity that cares about it. It is also the only entity that already interacts with the bootloader (ostree, during the installation phase) so it seems reasonable to keep all related logic in one place. Possible code duplication in (not yet existing) package managers is, IMO, outweighed by simplification of SotaUptaneClient code and class interfaces.

Reply:
The code duplication is just a minor issue here and I think it actually can be avoided - why not reuse Bootloader class in multiple package managers if it's required provided that there is a single/only instance of package manager in a scope of aktualizr/sotauptaneclient. But, if we move to approach where we can have more than one instance of the package manager and if two or more of them need the bootloading functionality then we should have a single instance of Bootloader shared across them (the same as we do with the only instance of Storage).

Another really important question here is to get rid of the bootloading specific methods out of the packagemanager interface (rebootDetected() and rebootFlagClear()). @eu-smirnov Does your vision on this matter imply it ?

Reply:
Absolutely, they shouldn't be a part of the package manager interface. This should be handled internally and only `finalizeInstall()` call exposed via the interface.

Reply:
Great ! So, we are on the same page on this specific item.

Now, another question, what would be the criterion for sotauptaneclient and aktualizr_secondary to invoke finalizeInstall() ? Is it storage->hasPendingInstall() ? I am afraid it won't always work as it can return true just after aktualizr instance/daemon restart without actual system reboot so finalizeInstall() will return an error which will eventually be sent to BE.

Reply:
Yes, it will return true after a simple service restart. But as I mentioned earlier, `finalizeInstall()` should be able to return three states: success, failure, and "try again" ( or "preconditions not met", or something like that). So in this particular case, we are calling `finalizeInstall()`, it internally checks `rebootDetected()` condition, which is false, and returns "try again" and we simply continue execution, just as we do now.

Reply:
Yeah, that is what I thought about too. So, it looks like we are on the same page on this matter too (I mean on the criterion of triggering finalizeInstall() and its return handling).

But, it does not necessary mean that an instantiation of Bootloader should happen within of a specific package manager constructor. What are drawbacks of passing a reference/pointer to Bootloader instance to a package manager ctor ?

Reply:
Drawbacks are that we complicate class structure and increase the amount of code. All package manager implementations would have to support taking Bootloader as a parameter even if they have nothing to do with it. I think in general it's better to instantiate/keep objects closer to the place they are being used. SotaUptaneClient has no use of Bootloader except for passing it to PackageManager.
I'm not even sure that Bootloader as a class should exist, to me, it looks more like a couple of helper functions in a header that PackageManagers can use if they need to.

Reply:
> All package manager implementations would have to support taking Bootloader as a parameter even if they have nothing to do with it.

Agree, I don't like it neither, moreover, I think the packagemanager interface's ctor shouldn't have any parameter at all, otherwise it (interface) imposes policies to its specializations (specific implementations), i.e. accept certain params in their ctors and have dependencies on stuff that it might not necessary need.

But, it does not contradict (or is not an obstacle) for passing a bootloader instance to a specific package manager ctor (the ostree pacman in this specific case). AFAIK, the best practice here is to pass references/pointers to an interface/base class in ctors to make sure we follow the 'D' principle of SOLID. In this case, we could avoid making assumptions in our tests about the bootloader implementations and doing something like this `boost::filesystem::remove(conf.bootloader.reboot_sentinel_dir / conf.bootloader.reboot_sentinel_name);` in order to emulate reboot. Instead those tests should just call bootloader->reboot(). Another advantage of it is the case when a customer would like to introduce its own package manager and to reuse most of the ostree pacman but have its own bootloading logic or vise versa a package manager that does not rely on the ostree pacman at all but it would like to reuse the existing bootloader instance.

Reply:
> Drawbacks are that we complicate class structure and increase the amount of code.

I disagree with this statement. If a bootloader instance is not passed to the package manager ctor then the later should include business logic of bootloader configuration parsing and the bootloader instantiation. It means more code within  the pacman scope. The best practice here is to have ctor as generic as possible and its parameters be abstract/interfaces/base_classes and have a separate standalone factory/builder that exposes a number of ways to instantiate an object and encapsulates a logic of the params instantiation. This design is more flexible and leaves a room for extension and reusability, decreases dependencies  and testability.

Having said that, I think, in this particular case, the bootloader case, it's not really critical as this really small piece of functionality, except perhaps the testing use-case when we make an assumption about the bootloader work internals, if it changes we have to update each test making this assumption.

Reply:
I'd still prefer to simplify class hierarchy and to reduce the amount of code now over having some abstractions for use cases we might never have. Even if we'll ever need that, I think that abstracting Bootloader from PackageManager could be done without SotaUptaneClient class knowing anything about it.

Reply:
>  think that abstracting Bootloader from PackageManager could be done without SotaUptaneClient class knowing anything about it.

what I am saying here is orthogonal to whether SotaUptaneClient knows or not about Bootloader.

>  to simplify class hierarchy and to reduce the amount of code now over having some abstractions for use cases we might never have

I am not proposing to introduce any new abstraction here, I am proposing passing of a bootloader instance to the ostree pacman ctor which actually slightly reduces the amount of its code.

Reply:
If I understand your proposal correctly, SotaUptaneClient will instantiate Bootloader and then instantiate PackageManager passing Bootloader to its constructor. In this approach, SotaUptaneClient is the owner of both instances and must make sure that PackageManager doesn't outlive Bootloader. In my opinion, this doesn't make sense as the PackageManager is the single user of Bootloader.

The abstraction I meant is that if we want to be consistent here, we should define a BootloaderInterface,  assuming that we might have different Bootloader implementations, and pass it as a parameter to PackageManager. This is the path I definitely don't want to take now.

Reply:
> If I understand your proposal correctly, SotaUptaneClient will instantiate Bootloader and then instantiate PackageManager passing Bootloader to its constructor

No, it's not what I think would be the best approach here. Let's discuss it tomorrow just after the standup.

> The abstraction I meant is that if we want to be consistent here, we should define a BootloaderInterface, assuming that we might have different Bootloader implementations, and pass it as a parameter to PackageManager. This is the path I definitely don't want to take now.

Agree, I think doesn't make sense to do it now, either.

Reply:
Sorry, I misunderstood it then. Let's discuss tomorrow.